src/zenml/utils/run_utils.py

```diff
@@ -72,3 +75,84 @@ def stop_run(run: PipelineRunResponse, graceful: bool = False) -> None:
 
     # Stop the run
     orchestrator.stop_run(run=run, graceful=graceful)
+
+
+def refresh_run_status(
+    run: PipelineRunResponse,
+    include_step_updates: bool = False,
+    zen_store: Optional["BaseZenStore"] = None,
+) -> PipelineRunResponse:
+    """Refresh the status of a pipeline run if it is initializing/running.
+
+    Args:
+        run: The pipeline run to refresh.
+        include_step_updates: Flag deciding whether we should also refresh
+            the status of individual steps.
+        zen_store: Optional ZenStore to use for updating the run. If not provided,
+            the ZenStore will be fetched from the Client. This is mainly useful
+            for running inside the ZenML server.
+
+    Returns:
+        The refreshed pipeline run.
+    """
+    # Check if the stack still accessible
+    if run.stack is None:
+        raise ValueError(
+            "The stack that this pipeline run response was executed on"
+            "is either not accessible or has been deleted."
+        )
+
+    # Check if the orchestrator is still accessible
+    orchestrator_list = run.stack.components.get(
+        StackComponentType.ORCHESTRATOR, []
+    )
+    if len(orchestrator_list) == 0:
+        raise ValueError(
+            "The orchestrator that this pipeline run response was "
+            "executed with is either not accessible or has been deleted."
+        )
+
+    orchestrator = cast(
+        BaseOrchestrator,
+        StackComponent.from_model(component_model=orchestrator_list[0]),
+    )
+
+    # Fetch the status (and optionally step statuses)
+    pipeline_status, step_statuses = orchestrator.fetch_status(
+        run=run, include_steps=include_step_updates
+    )
+
+    if zen_store is None:
+        from zenml.client import Client
+
+        zen_store = Client().zen_store
+    else:
+        zen_store = zen_store
+
+    # Update step statuses
+    if include_step_updates:
+        if step_statuses:
+            current_steps = run.steps
+            for step_name, step_status in step_statuses.items():
+                if current_steps[step_name].status.is_finished:
```

Comment:
`current_steps` might not always contain that step. For example: The job already exists, which means your logic in the KubernetesOrchestrator returns something for that step. But the job has not yet reached the point in code where it registered the step in the server yet. So in that case, we do not want to update anything as that will fail.

Reply:
Added an additional check. If the `step_name` is not in the run.steps yet, we skip.

Reply:
Good point, added it.